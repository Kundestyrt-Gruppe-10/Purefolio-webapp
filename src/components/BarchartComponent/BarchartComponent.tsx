import React from 'react';
import {
  BarChart,
  Bar,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
} from 'recharts';
import { NaceRegion, NaceRegionData, EuroStatTable } from '../../types';
import { handleColorType } from '../NaceRegionCard/NaceRegionCard';
// TODO: Add tests

/**
 * Input: [
 * [{
 *  nace=,
 *  Region=
 *  Year
 *  Emission
 * ,
 *  nace=,
 *  Region=
 *  Year
 *  Emission
 * ]
 * ]
 * Output: [
 *  {
 *  Year:
 *  Emmision1
 *  Emission2
 *  Emission3
 * }
 * ]
 */

interface Props {
  naceRegionData: NaceRegionData[][];
  naceRegionList: NaceRegion[];
  //esgFactorInfo: EuroStatTable;
  esgFactor:
    | 'emissionPerYear'
    | 'workAccidentsIncidentRate'
    | 'genderPayGap'
    | 'environmentTaxes'
    | 'fatalAccidentsAtWork'
    | 'temporaryemployment'
    | 'employeesPrimaryEducation'
    | 'employeesSecondaryEducation'
    | 'employeesTertiaryEducation';
}

interface NaceRegionChartItem {
  year: number; // year
  [dataKey: string]: number | undefined; // naceRegionKey: data
}
export const BarchartComponent: React.FC<Props> = ({
  naceRegionData,
  naceRegionList,
  esgFactor,
}) => {
  const naceRegionItems: NaceRegionChartItem[] = [];

  // TODO: This should be a function and moved out from component
  naceRegionData.forEach((naceRegion: NaceRegionData[]) => {
    naceRegion.forEach((element: NaceRegionData) => {
      // If year already in naceRegionItems, update existing object
      naceRegionItems.map(
        () =>
          // Return object if it exist with new key:value pair
          // or create new object if it does not exist
          ((naceRegionItems.find((el) => el.year === element.year) || {
            year: element.year,
          })[element.region.regionName + element.nace.naceCode] =
            element[esgFactor]),
      );
      // Runs when element not yet in naceRegionItems
      const found = naceRegionItems.some((el) => el.year === element.year);
      if (!found) {
        naceRegionItems.push({
          year: element.year,
          [element.region.regionName + element.nace.naceCode]: element[
            esgFactor
          ],
        });
      }
    });
  });
  return (
    <>
      <BarChart
        width={1000}
        height={500}
        data={naceRegionItems}
        margin={{
          top: 5,
          right: 30,
          left: 20,
          bottom: 5,
        }}
      >
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="year" />
        <YAxis />
        <Tooltip />
        <Legend />
        {naceRegionList.map((item, idx) => {
          return (
            // This is number of Bars per group/ year. If we compare multiple NaceRegions
            // We will have multiple bars
            <Bar
              data-testid="bar"
              key={idx}
              dataKey={item.region.regionName + item.nace.naceCode}
              fill={handleColorType(idx)}
            />
          );
        })}
      </BarChart>
    </>
  );
};
