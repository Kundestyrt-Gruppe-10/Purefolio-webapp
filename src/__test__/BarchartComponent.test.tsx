import { render } from '@testing-library/react';
import React from 'react';
import { BarchartComponent } from '../components/BarchartComponent/BarchartComponent';
import {
  naceRegionDataListList_1_example,
  naceRegionDataListList_3_examples,
  naceRegionList_1,
  naceRegionList_3,
  // Arrange
} from '../mockData';

describe('<BarchartComponent />', () => {
  it('Matches snapshot when 1 NaceCard is chosen', () => {
    // Act
    const documentBody = render(
      <BarchartComponent
        naceRegionData={naceRegionDataListList_1_example}
        esgFactor="emissionPerYear"
        naceRegionList={naceRegionList_1}
      />,
    );
    // Assert
    expect(documentBody).toMatchSnapshot();
  });

  it('Matches snapshot when 3 NaceCard is chosen', () => {
    // Act
    const documentBody = render(
      <BarchartComponent
        naceRegionData={naceRegionDataListList_3_examples}
        esgFactor="emissionPerYear"
        naceRegionList={naceRegionList_3}
      />,
    );
    // Assert
    expect(documentBody).toMatchSnapshot();
  });
  // Assert
});
