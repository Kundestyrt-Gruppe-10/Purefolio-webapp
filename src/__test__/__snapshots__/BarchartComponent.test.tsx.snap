// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BarchartComponent /> Matches snapshot when 1 NaceCard is chosen 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gKAblj pSuCS"
      >
        <div
          class="sc-iCoHVE dWRZZO"
        >
          <div
            class="sc-fujyUd VeEOy"
          >
            <div
              class="recharts-responsive-container"
              style="width: 97%; height: 97%;"
            >
              <div
                style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
              />
            </div>
          </div>
          <div
            class="sc-pNWxx fxawGb"
          >
            <div
              class="sc-jrsJCI fzuYIX"
            >
              <div
                class="sc-kEqYlL fjxjkT"
              >
                History Graph
              </div>
              <div
                class="sc-iqAbSa jdvqLV"
              >
                Gram per Euro
              </div>
            </div>
            <div
              class="sc-crzoUp iKdPGR"
            >
              <div
                class="sc-dIsAE hOSeLT"
              >
                <div
                  class="sc-ksluoS gMpDis"
                >
                  ESG Factor:
                </div>
                <div
                  class="sc-hBMVcZ iYOQhu"
                >
                  Emmisions of Greenhouse gases
                </div>
              </div>
              <div
                class="sc-bqGHjH eDezpz"
              >
                <div
                  class="sc-ksluoS gMpDis"
                >
                  Year:
                </div>
                <div
                  class="sc-hBMVcZ iYOQhu"
                >
                   
                  2015
                   - 
                  2018
                </div>
              </div>
            </div>
            <div
              class="sc-fnVYJo nUgzX"
            >
              Air Emissions Accounts record the flows of residual gaseous and particulate materials emitted by resident units and flowing into the atmosphere. Residual gaseous and particulate materials are the physical flows of gaseous or particulate materials (‘air emissions’).
              <div
                class="sc-fFSRdu gYZDfv"
              >
                <a
                  class="sc-bkbjAj"
                  href="https://ec.europa.eu/eurostat/cache/metadata/en/env_ac_ainah_r2_esms.htm"
                >
                  https://ec.europa.eu/eurostat/cache/metadata/en/env_ac_ainah_r2_esms.htm
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gKAblj pSuCS"
    >
      <div
        class="sc-iCoHVE dWRZZO"
      >
        <div
          class="sc-fujyUd VeEOy"
        >
          <div
            class="recharts-responsive-container"
            style="width: 97%; height: 97%;"
          >
            <div
              style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
            />
          </div>
        </div>
        <div
          class="sc-pNWxx fxawGb"
        >
          <div
            class="sc-jrsJCI fzuYIX"
          >
            <div
              class="sc-kEqYlL fjxjkT"
            >
              History Graph
            </div>
            <div
              class="sc-iqAbSa jdvqLV"
            >
              Gram per Euro
            </div>
          </div>
          <div
            class="sc-crzoUp iKdPGR"
          >
            <div
              class="sc-dIsAE hOSeLT"
            >
              <div
                class="sc-ksluoS gMpDis"
              >
                ESG Factor:
              </div>
              <div
                class="sc-hBMVcZ iYOQhu"
              >
                Emmisions of Greenhouse gases
              </div>
            </div>
            <div
              class="sc-bqGHjH eDezpz"
            >
              <div
                class="sc-ksluoS gMpDis"
              >
                Year:
              </div>
              <div
                class="sc-hBMVcZ iYOQhu"
              >
                 
                2015
                 - 
                2018
              </div>
            </div>
          </div>
          <div
            class="sc-fnVYJo nUgzX"
          >
            Air Emissions Accounts record the flows of residual gaseous and particulate materials emitted by resident units and flowing into the atmosphere. Residual gaseous and particulate materials are the physical flows of gaseous or particulate materials (‘air emissions’).
            <div
              class="sc-fFSRdu gYZDfv"
            >
              <a
                class="sc-bkbjAj"
                href="https://ec.europa.eu/eurostat/cache/metadata/en/env_ac_ainah_r2_esms.htm"
              >
                https://ec.europa.eu/eurostat/cache/metadata/en/env_ac_ainah_r2_esms.htm
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<BarchartComponent /> Matches snapshot when 3 NaceCard is chosen 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gKAblj pSuCS"
      >
        <div
          class="sc-iCoHVE dWRZZO"
        >
          <div
            class="sc-fujyUd VeEOy"
          >
            <div
              class="recharts-responsive-container"
              style="width: 97%; height: 97%;"
            >
              <div
                style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
              />
            </div>
          </div>
          <div
            class="sc-pNWxx fxawGb"
          >
            <div
              class="sc-jrsJCI fzuYIX"
            >
              <div
                class="sc-kEqYlL fjxjkT"
              >
                History Graph
              </div>
              <div
                class="sc-iqAbSa jdvqLV"
              >
                Gram per Euro
              </div>
            </div>
            <div
              class="sc-crzoUp iKdPGR"
            >
              <div
                class="sc-dIsAE hOSeLT"
              >
                <div
                  class="sc-ksluoS gMpDis"
                >
                  ESG Factor:
                </div>
                <div
                  class="sc-hBMVcZ iYOQhu"
                >
                  Emmisions of Greenhouse gases
                </div>
              </div>
              <div
                class="sc-bqGHjH eDezpz"
              >
                <div
                  class="sc-ksluoS gMpDis"
                >
                  Year:
                </div>
                <div
                  class="sc-hBMVcZ iYOQhu"
                >
                   
                  2015
                   - 
                  2018
                </div>
              </div>
            </div>
            <div
              class="sc-fnVYJo nUgzX"
            >
              Air Emissions Accounts record the flows of residual gaseous and particulate materials emitted by resident units and flowing into the atmosphere. Residual gaseous and particulate materials are the physical flows of gaseous or particulate materials (‘air emissions’).
              <div
                class="sc-fFSRdu gYZDfv"
              >
                <a
                  class="sc-bkbjAj"
                  href="https://ec.europa.eu/eurostat/cache/metadata/en/env_ac_ainah_r2_esms.htm"
                >
                  https://ec.europa.eu/eurostat/cache/metadata/en/env_ac_ainah_r2_esms.htm
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gKAblj pSuCS"
    >
      <div
        class="sc-iCoHVE dWRZZO"
      >
        <div
          class="sc-fujyUd VeEOy"
        >
          <div
            class="recharts-responsive-container"
            style="width: 97%; height: 97%;"
          >
            <div
              style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
            />
          </div>
        </div>
        <div
          class="sc-pNWxx fxawGb"
        >
          <div
            class="sc-jrsJCI fzuYIX"
          >
            <div
              class="sc-kEqYlL fjxjkT"
            >
              History Graph
            </div>
            <div
              class="sc-iqAbSa jdvqLV"
            >
              Gram per Euro
            </div>
          </div>
          <div
            class="sc-crzoUp iKdPGR"
          >
            <div
              class="sc-dIsAE hOSeLT"
            >
              <div
                class="sc-ksluoS gMpDis"
              >
                ESG Factor:
              </div>
              <div
                class="sc-hBMVcZ iYOQhu"
              >
                Emmisions of Greenhouse gases
              </div>
            </div>
            <div
              class="sc-bqGHjH eDezpz"
            >
              <div
                class="sc-ksluoS gMpDis"
              >
                Year:
              </div>
              <div
                class="sc-hBMVcZ iYOQhu"
              >
                 
                2015
                 - 
                2018
              </div>
            </div>
          </div>
          <div
            class="sc-fnVYJo nUgzX"
          >
            Air Emissions Accounts record the flows of residual gaseous and particulate materials emitted by resident units and flowing into the atmosphere. Residual gaseous and particulate materials are the physical flows of gaseous or particulate materials (‘air emissions’).
            <div
              class="sc-fFSRdu gYZDfv"
            >
              <a
                class="sc-bkbjAj"
                href="https://ec.europa.eu/eurostat/cache/metadata/en/env_ac_ainah_r2_esms.htm"
              >
                https://ec.europa.eu/eurostat/cache/metadata/en/env_ac_ainah_r2_esms.htm
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
